File: find_problem_folder.js
#!/usr/bin/env node
/**
 * Fetch and analyze Yandex Disk folders at offset 400 to find VARCHAR violations.
 * Run this script: node find_problem_folder.js
 */

const https = require('https');
const fs = require('fs');

function extractWorkInfo(folderName) {
    const match = folderName.trim().match(/^(.+?)\s*\((.+?)\)\s*$/);
    if (match) {
        return {
            title: match[1].trim(),
            work_type: match[2].trim()
        };
    }
    return {
        title: folderName,
        work_type: 'неизвестный тип'
    };
}

function determineSubject(title) {
    const t = title.toLowerCase();
    
    if (/электро|электри|энергет|эу|ру/.test(t)) return 'электроэнергетика';
    if (/автоматиз|управлен|асу|контрол|регулир/.test(t)) return 'автоматизация';
    if (/строител|бетон|конструк|здание|сооружен/.test(t)) return 'строительство';
    if (/механ|привод|станок|оборудован/.test(t)) return 'механика';
    if (/газ|газопровод|нефт/.test(t)) return 'газоснабжение';
    if (/програм|по|software|алгоритм|дискрет/.test(t)) return 'программирование';
    if (/безопасн|охран|труд|защит/.test(t)) return 'безопасность';
    if (/тепло|водоснабжен|вентиляц|отоплен/.test(t)) return 'теплоснабжение';
    if (/транспорт|дорог|судов|автомобил|локомотив/.test(t)) return 'транспорт';
    if (/гидравлик|гидро/.test(t)) return 'гидравлика';
    
    return 'общая инженерия';
}

const publicKey = encodeURIComponent('https://disk.yandex.ru/d/usjmeUqnkY9IfQ');
const url = `https://cloud-api.yandex.net/v1/disk/public/resources?public_key=${publicKey}&limit=100&offset=400`;

console.log('Fetching data from Yandex Disk API at offset 400...');
console.log('='.repeat(80));

https.get(url, (res) => {
    let data = '';
    
    res.on('data', (chunk) => {
        data += chunk;
    });
    
    res.on('end', () => {
        try {
            const response = JSON.parse(data);
            
            // Save raw response
            fs.writeFileSync('offset_400_raw.json', JSON.stringify(response, null, 2));
            console.log('✓ Raw data saved to offset_400_raw.json\n');
            
            if (!response._embedded || !response._embedded.items) {
                console.log('ERROR: No items found in API response');
                return;
            }
            
            const items = response._embedded.items;
            console.log(`Found ${items.length} items\n`);
            
            const TITLE_LIMIT = 1000;
            const WORK_TYPE_LIMIT = 100;
            const SUBJECT_LIMIT = 200;
            
            const violations = [];
            const allItems = [];
            
            items.forEach((item, i) => {
                if (item.type === 'dir') {
                    const folderName = item.name || '';
                    const workInfo = extractWorkInfo(folderName);
                    
                    const title = workInfo.title;
                    const workType = workInfo.work_type;
                    const subject = determineSubject(title);
                    
                    const itemData = {
                        offset_index: 400 + i,
                        folder_name: folderName,
                        title: title,
                        title_length: title.length,
                        work_type: workType,
                        work_type_length: workType.length,
                        subject: subject,
                        subject_length: subject.length,
                        violations: []
                    };
                    
                    if (title.length > TITLE_LIMIT) {
                        itemData.violations.push({
                            field: 'title',
                            length: title.length,
                            limit: TITLE_LIMIT,
                            exceeds_by: title.length - TITLE_LIMIT
                        });
                        violations.push(itemData);
                    }
                    
                    if (workType.length > WORK_TYPE_LIMIT) {
                        itemData.violations.push({
                            field: 'work_type',
                            length: workType.length,
                            limit: WORK_TYPE_LIMIT,
                            exceeds_by: workType.length - WORK_TYPE_LIMIT
                        });
                        if (!violations.includes(itemData)) {
                            violations.push(itemData);
                        }
                    }
                    
                    if (subject.length > SUBJECT_LIMIT) {
                        itemData.violations.push({
                            field: 'subject',
                            length: subject.length,
                            limit: SUBJECT_LIMIT,
                            exceeds_by: subject.length - SUBJECT_LIMIT
                        });
                        if (!violations.includes(itemData)) {
                            violations.push(itemData);
                        }
                    }
                    
                    allItems.push(itemData);
                }
            });
            
            // Save analysis
            const analysis = {
                total_items: allItems.length,
                violations_count: violations.length,
                violations: violations,
                all_items: allItems
            };
            
            fs.writeFileSync('offset_400_analysis.json', JSON.stringify(analysis, null, 2));
            console.log('✓ Analysis saved to offset_400_analysis.json\n');
            
            // Print summary
            if (violations.length > 0) {
                console.log('='.repeat(80));
                console.log(`FOUND ${violations.length} VIOLATION(S)!`);
                console.log('='.repeat(80));
                
                violations.forEach(item => {
                    console.log('\n' + '='.repeat(80));
                    console.log(`Offset Index: ${item.offset_index}`);
                    console.log('='.repeat(80));
                    console.log(`Folder Name: ${item.folder_name}`);
                    console.log(`\nExtracted Fields:`);
                    console.log(`  Title (${item.title_length} chars): ${item.title.substring(0, 100)}...`);
                    console.log(`  Work Type (${item.work_type_length} chars): ${item.work_type.substring(0, 100)}...`);
                    console.log(`  Subject (${item.subject_length} chars): ${item.subject}`);
                    console.log(`\nViolations:`);
                    item.violations.forEach(v => {
                        console.log(`  ❌ ${v.field.toUpperCase()}: ${v.length} chars (limit: ${v.limit}, exceeds by ${v.exceeds_by})`);
                    });
                });
            } else {
                console.log('='.repeat(80));
                console.log('No violations found!');
                console.log('='.repeat(80));
                console.log('\nShowing top 10 longest titles:');
                allItems.sort((a, b) => b.title_length - a.title_length).slice(0, 10).forEach((item, i) => {
                    console.log(`  ${i+1}. Offset ${item.offset_index}: ${item.title_length} chars`);
                    console.log(`     ${item.title.substring(0, 80)}...`);
                });
                
                console.log('\nShowing top 10 longest work types:');
                allItems.sort((a, b) => b.work_type_length - a.work_type_length).slice(0, 10).forEach((item, i) => {
                    console.log(`  ${i+1}. Offset ${item.offset_index}: ${item.work_type_length} chars`);
                    console.log(`     ${item.work_type.substring(0, 80)}...`);
                });
            }
            
            console.log('\n' + '='.repeat(80));
            console.log('Analysis complete!');
            console.log('='.repeat(80));
            
        } catch (error) {
            console.error('Error parsing response:', error);
            console.error('Raw response:', data.substring(0, 500));
        }
    });
}).on('error', (error) => {
    console.error('Error fetching data:', error);
});

File: find_problem_folder.py
#!/usr/bin/env python3
"""
Fetch and analyze Yandex Disk folders at offset 400 to find VARCHAR violations.
Run this script locally: python3 find_problem_folder.py
"""

import json
import urllib.request
import urllib.parse
import re

def extract_work_info(folder_name: str):
    """Extract title and work type from folder name"""
    match = re.match(r'^(.+?)\s*\((.+?)\)\s*$', folder_name.strip())
    if match:
        return {
            'title': match.group(1).strip(),
            'work_type': match.group(2).strip()
        }
    return {
        'title': folder_name,
        'work_type': 'неизвестный тип'
    }

def determine_subject(title: str) -> str:
    """Determine subject area from title keywords"""
    t = title.lower()
    
    if re.search(r'электро|электри|энергет|эу|ру', t):
        return 'электроэнергетика'
    if re.search(r'автоматиз|управлен|асу|контрол|регулир', t):
        return 'автоматизация'
    if re.search(r'строител|бетон|конструк|здание|сооружен', t):
        return 'строительство'
    if re.search(r'механ|привод|станок|оборудован', t):
        return 'механика'
    if re.search(r'газ|газопровод|нефт', t):
        return 'газоснабжение'
    if re.search(r'програм|по|software|алгоритм|дискрет', t):
        return 'программирование'
    if re.search(r'безопасн|охран|труд|защит', t):
        return 'безопасность'
    if re.search(r'тепло|водоснабжен|вентиляц|отоплен', t):
        return 'теплоснабжение'
    if re.search(r'транспорт|дорог|судов|автомобил|локомотив', t):
        return 'транспорт'
    if re.search(r'гидравлик|гидро', t):
        return 'гидравлика'
    
    return 'общая инженерия'

def main():
    print("Fetching data from Yandex Disk API at offset 400...")
    print("="*80)
    
    # Fetch data
    public_key = 'https://disk.yandex.ru/d/usjmeUqnkY9IfQ'
    api_url = f'https://cloud-api.yandex.net/v1/disk/public/resources?public_key={urllib.parse.quote(public_key)}&limit=100&offset=400'
    
    try:
        req = urllib.request.Request(api_url)
        with urllib.request.urlopen(req, timeout=30) as response:
            data = json.loads(response.read().decode('utf-8'))
        
        # Save raw response
        with open('offset_400_raw.json', 'w', encoding='utf-8') as f:
            json.dump(data, f, ensure_ascii=False, indent=2)
        print("✓ Raw data saved to offset_400_raw.json\n")
        
        if '_embedded' not in data or 'items' not in data['_embedded']:
            print("ERROR: No items found in API response")
            return
        
        items = data['_embedded']['items']
        print(f"Found {len(items)} items\n")
        
        # Field limits
        TITLE_LIMIT = 1000
        WORK_TYPE_LIMIT = 100
        SUBJECT_LIMIT = 200
        
        violations = []
        all_items = []
        
        # Analyze each folder
        for i, item in enumerate(items):
            if item.get('type') == 'dir':
                folder_name = item.get('name', '')
                work_info = extract_work_info(folder_name)
                
                title = work_info['title']
                work_type = work_info['work_type']
                subject = determine_subject(title)
                
                item_data = {
                    'offset_index': 400 + i,
                    'folder_name': folder_name,
                    'title': title,
                    'title_length': len(title),
                    'work_type': work_type,
                    'work_type_length': len(work_type),
                    'subject': subject,
                    'subject_length': len(subject),
                    'violations': []
                }
                
                # Check for violations
                if len(title) > TITLE_LIMIT:
                    item_data['violations'].append({
                        'field': 'title',
                        'length': len(title),
                        'limit': TITLE_LIMIT,
                        'exceeds_by': len(title) - TITLE_LIMIT
                    })
                    violations.append(item_data)
                
                if len(work_type) > WORK_TYPE_LIMIT:
                    item_data['violations'].append({
                        'field': 'work_type',
                        'length': len(work_type),
                        'limit': WORK_TYPE_LIMIT,
                        'exceeds_by': len(work_type) - WORK_TYPE_LIMIT
                    })
                    if item_data not in violations:
                        violations.append(item_data)
                
                if len(subject) > SUBJECT_LIMIT:
                    item_data['violations'].append({
                        'field': 'subject',
                        'length': len(subject),
                        'limit': SUBJECT_LIMIT,
                        'exceeds_by': len(subject) - SUBJECT_LIMIT
                    })
                    if item_data not in violations:
                        violations.append(item_data)
                
                all_items.append(item_data)
        
        # Save analysis
        analysis = {
            'total_items': len(all_items),
            'violations_count': len(violations),
            'violations': violations,
            'all_items': all_items
        }
        
        with open('offset_400_analysis.json', 'w', encoding='utf-8') as f:
            json.dump(analysis, f, ensure_ascii=False, indent=2)
        print("✓ Analysis saved to offset_400_analysis.json\n")
        
        # Print summary
        if violations:
            print("="*80)
            print(f"FOUND {len(violations)} VIOLATION(S)!")
            print("="*80)
            
            for item in violations:
                print(f"\n{'='*80}")
                print(f"Offset Index: {item['offset_index']}")
                print(f"{'='*80}")
                print(f"Folder Name: {item['folder_name']}")
                print(f"\nExtracted Fields:")
                print(f"  Title ({item['title_length']} chars): {item['title'][:100]}...")
                print(f"  Work Type ({item['work_type_length']} chars): {item['work_type'][:100]}...")
                print(f"  Subject ({item['subject_length']} chars): {item['subject']}")
                print(f"\nViolations:")
                for v in item['violations']:
                    print(f"  ❌ {v['field'].upper()}: {v['length']} chars (limit: {v['limit']}, exceeds by {v['exceeds_by']})")
                print()
        else:
            print("="*80)
            print("No violations found!")
            print("="*80)
            print("\nShowing top 10 longest titles:")
            sorted_by_title = sorted(all_items, key=lambda x: x['title_length'], reverse=True)
            for i, item in enumerate(sorted_by_title[:10]):
                print(f"  {i+1}. Offset {item['offset_index']}: {item['title_length']} chars")
                print(f"     {item['title'][:80]}...")
            
            print("\nShowing top 10 longest work types:")
            sorted_by_work_type = sorted(all_items, key=lambda x: x['work_type_length'], reverse=True)
            for i, item in enumerate(sorted_by_work_type[:10]):
                print(f"  {i+1}. Offset {item['offset_index']}: {item['work_type_length']} chars")
                print(f"     {item['work_type'][:80]}...")
        
        print("\n" + "="*80)
        print("Analysis complete!")
        print("="*80)
        
    except Exception as e:
        print(f"ERROR: {e}")
        import traceback
        traceback.print_exc()

if __name__ == '__main__':
    main()
